Sources/HummingbirdCore/Server/HTTP/HTTPChannelHandler.swift

```diff
@@ -33,83 +33,77 @@ enum HTTPChannelError: Error {
+                            }
+                            do {
+                                try await outbound.write(.head(response.head))
+                                let tailHeaders = try await response.body.write(responseWriter)
+                                try await outbound.write(.end(tailHeaders))
+                            } catch {
+                                throw error
+                            }
+                            if request.headers[.connection] == "close" {
+                                throw HTTPChannelError.closeConnection
+                            }
+
+                            // Flush current request
+                            // read until we don't have a body part
+                            var part: HTTPRequestPart?
+                            while true {
+                                part = try await iterator.next()
+                                guard case .body = part else { break }
+                            }
+                            // if we have an end then read the next part
+                            if case .end = part {
+                                part = try await iterator.next()
+                            }
+
+                            // if part is nil break out of loop
+                            guard let part else {
+                                break
+                            }
+
+                            // Process next request unless it is cancelled then exit
+                            if Task.isShuttingDownGracefully || Task.isCancelled {
+                                return
+                            }
+
+                            // part should be a head, if not throw error
+                            guard case .head(let newHead) = part else { throw HTTPChannelError.unexpectedHTTPPart(part) }
+                            head = newHead
                         }
-                        // if we have an end then read the next part
-                        if case .end = part {
-                            part = try await iterator.next()
-                        }
-
-                        // if part is nil break out of loop
-                        guard let part else {
-                            break
-                        }
-
-                        // part should be a head, if not throw error
-                        guard case .head(let newHead) = part else { throw HTTPChannelError.unexpectedHTTPPart(part) }
-                        head = newHead
                     }
                 }
-            } onGracefulShutdown: {
-                // set to cancelled
-                if processingRequest.exchange(.cancelled, ordering: .relaxed) == .idle {
-                    // only close the channel input if it is idle
-                    asyncChannel.channel.close(mode: .input, promise: nil)
-                }
+
+                try await asyncChannel.channel.closeFuture.get()
```

Comment:
This shouldn't be needed. `executeThenClose` makes sure that the channel is closed

Reply:
The reason I do this is because I want to trigger `.cancelAll()` if the channel closes while a request is still being handled